inc/appliance.class.php

```diff
@@ -0,0 +1,626 @@
+               'itemtype_name' => 'typeitem',
+               'itemtypes'     => self::getTypes(true),
+               'checkright'    => true
+            ]);
+            echo Html::submit(_x('button', 'Post'), ['name' => 'massiveaction']);
+            return true;
+
+         case "transfer" :
+            Dropdown::show('Entity');
+            echo Html::submit(_x('button', 'Post'), ['name' => 'massiveaction']);
+            return true;
+
+      }
+      return parent::showMassiveActionsSubForm($ma);
+   }
+
+
+   static function processMassiveActionsForOneItemtype(MassiveAction $ma, CommonDBTM $item, array $ids) {
+
+      $appliance_item = new Appliance_Item();
+
+      switch ($ma->getAction()) {
+         case "add_item":
+            $input = $ma->getInput();
+            foreach ($ids as $id) {
+               $input = [
+                  'appliances_id'   => $input['appliances_id'],
+                  'items_id'        => $id,
+                  'itemtype'        => $item->getType()
+               ];
+               if ($appliance_item->can(-1, CREATE, $input)) {
+                  if ($appliance_item->add($input)) {
+                     $ma->itemDone($item->getType(), $id, MassiveAction::ACTION_OK);
+                  } else {
+                     $ma->itemDone($item->getType(), $ids, MassiveAction::ACTION_KO);
+                  }
+               } else {
+                  $ma->itemDone($item->getType(), $ids, MassiveAction::ACTION_KO);
+               }
+            }
+            return;
+
+         case "transfer" :
+            $input = $ma->getInput();
+            if ($item->getType() == 'Appliance') {
+               foreach ($ids as $key) {
+                  $item->getFromDB($key);
+                  $type = ApplianceType::transfer($item->fields["appliancetypes_id"], $input['entities_id']);
+                  if ($type > 0) {
+                     $values["id"] = $key;
+                     $values["appliancetypes_id"] = $type;
+                     $item->update($values);
+                  }
+
+                     unset($values);
+                     $values["id"] = $key;
+                     $values["entities_id"] = $input['entities_id'];
+
+                  if ($item->update($values)) {
+                     $ma->itemDone($item->getType(), $key, MassiveAction::ACTION_OK);
+                  } else {
+                     $ma->itemDone($item->getType(), $key, MassiveAction::ACTION_KO);
+                  }
+               }
+            }
+            return;
+
+         case 'install' :
+            $input = $ma->getInput();
+            foreach ($ids as $key) {
+               if ($item->can($key, UPDATE)) {
+                  $values = [
+                     'appliances_id'   => $key,
+                     'items_id'        => $input["item_item"],
+                     'itemtype'        => $input['typeitem']
+                  ];
+                  if ($appliance_item->add($values)) {
```

Comment:
Add should be skipped if item is already associated to this appliance.
If a selected item is already associated, this error is trigerred:
```
glpisqllog.ERROR: DBmysql::query() in /var/www/glpi/inc/dbmysql.class.php line 305
  *** MySQL query error:
  SQL: INSERT INTO `glpi_appliances_items` (`appliances_id`, `items_id`, `itemtype`) VALUES ('1', '1', 'Computer')
  Error: Duplicate entry '1-1-Computer' for key 'unicity'
  Backtrace :
  inc/dbmysql.class.php:1131                         
  inc/commondbtm.class.php:592                       DBmysql->insert()
  inc/commondbtm.class.php:1129                      CommonDBTM->addToDB()
  inc/appliance.class.php:607                        CommonDBTM->add()
  inc/massiveaction.class.php:1050                   Appliance::processMassiveActionsForOneItemtype()
  inc/massiveaction.class.php:1029                   MassiveAction->processForSeveralItemtypes()
  front/massiveaction.php:59                         MassiveAction->process()
  {"user":"2@e906a4c5355e"} 
```